objects/src/accounts/delta/vault.rs

```diff
@@ -1,236 +1,488 @@
-                added.push(Asset::Fungible(asset));
-            } else {
-                removed.push(Asset::Fungible(asset));
-            }
+    /// Checks whether this vault delta is valid.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// - The number updated assets is greater than [u16::MAX].
+    /// - One or more fungible assets' faucet IDs are invalid.
+    fn validate(&self) -> Result<(), AccountDeltaError> {
+        if self.0.len() > u16::MAX as usize {
+            return Err(AccountDeltaError::TooManyFungibleAssets {
+                actual: self.0.len(),
+                max: u16::MAX as usize,
+            });
         }
 
-        for (non_fungible, is_added) in non_fungibles {
-            let asset = Asset::NonFungible(non_fungible);
-            if is_added {
-                added.push(asset);
-            } else {
-                removed.push(asset);
+        for faucet_id in self.0.keys() {
+            if !matches!(faucet_id.account_type(), AccountType::FungibleFaucet) {
+                return Err(AccountDeltaError::NotAFungibleFaucetId(*faucet_id));
             }
         }
 
-        let delta = Self::from_iterators(added, removed);
+        Ok(())
+    }
+}
+
+impl Serializable for FungibleAssetDelta {
+    fn write_into<W: ByteWriter>(&self, target: &mut W) {
+        target.write_u16(self.0.len().try_into().expect("too many fungible asset updates"));
+        // TODO: We save `i64` as `u64` since winter utils only support unsigned integers for now.
+        //   We should update this code (and deserialization as well) once it support signed
+        //   integers.
+        target.write_many(self.0.iter().map(|(&faucet_id, &delta)| (faucet_id, delta as u64)));
+    }
+}
+
+impl Deserializable for FungibleAssetDelta {
+    fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
+        let num_fungible_assets = source.read_u16()? as usize;
+        // TODO: We save `i64` as `u64` since winter utils only support unsigned integers for now.
+        //   We should update this code (and serialization as well) once it support signed integers.
+        let map = source
+            .read_many::<(AccountId, u64)>(num_fungible_assets)?
+            .into_iter()
+            .map(|(account_id, delta_as_u64)| (account_id, delta_as_u64 as i64))
+            .collect();
+
+        Self::new(map).map_err(|err| DeserializationError::InvalidValue(err.to_string()))
+    }
+}
+
+/// A binary tree map of non-fungible asset changes (addition and removal) in the account vault.
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+pub struct NonFungibleAssetDelta(BTreeMap<Digest, NonFungibleDeltaAction>);
+
+impl NonFungibleAssetDelta {
+    /// Validates and creates a new non-fungible asset delta.
+    ///
+    /// # Errors
+    /// Returns an error if the delta does not pass the validation.
+    pub fn new(map: BTreeMap<Digest, NonFungibleDeltaAction>) -> Result<Self, AccountDeltaError> {
```

Comment:
I think we should be able to use `BTreeMap<NonFungibleAsset, NonFungibleDeltaAction>` and then we may avoid the need for validation here.